### PR TITLE
Add a sorted set for replies-to-this-post

### PR DESCRIPTION
**Pull request by BenLubar**

### CODE REVIEW

**Comment by barisusakli** [comment]
This is similar to https://github.com/NodeBB/nodebb-plugin-threaded-replies, we even used the same sorted set name

I am not against merging this into core although we need to add support for it in the default themes. There needs to be some UI to expand the replies on each post.

Also I don't think we should remove/add on post delete, it makes more sense to just remove the post from `pid:<postData.toPid>:replies` when the post is purged.

Another thing the plugin handles is post move, when a post is moved it is also removed from the replies sorted set.

**Comment by BenLubar** [comment]
> There needs to be some UI to expand the replies on each post.

I was thinking something like "2 replies" right next to where the "in reply to" shows up (at least in Persona - I haven't used the other themes), and if you click on it, you get a dialog similar to the "fork topic" dialog but with a list of links that let you go to the post and its replies. (But I want feedback on that idea before I implement it.)

> Another thing the plugin handles is post move, when a post is moved it is also removed from the replies sorted set.

Replies still work when made in a different topic [(example)](https://what.thedailywtf.com/post/983704), so I'd rather keep them in the sorted set for moves.

> Also I don't think we should remove/add on post delete, it makes more sense to just remove the post from `pid:<postData.toPid>:replies` when the post is purged.

It appears that replies still show up on deleted posts, so that's a good point. https://community.nodebb.org/post/37043

**Comment by barisusakli** [comment]
I think for the UI it would be better if we just have a clickable "X replies" box that expands/collapses the replies inline under the post.

```
Post A
   + 2 Replies
```

When clicked

```
Post A
   - 2 replies
    Post B
    Post C
```

Pretty sure @psychobunny will have some ideas as well.

Delete posts are just hidden but still visible to admins/mods so we need them in the sets, they should be removed when purged.

**Comment by BenLubar** [comment]
> Delete posts are just hidden but still visible to admins/mods so we need them in the sets, they should be removed when purged.

Yeah, I was thinking the "in reply to" thing was hidden on deleted posts, but it's visible, so there's no extra information being revealed on deleted posts. I updated the PR with the change from deletion to purging.

File: src/posts/create.js [hunk 1]
```diff
@@ -86,6 +86,12 @@ module.exports = function(Posts) {
 						db.sortedSetAdd('posts:pid', timestamp, postData.pid, next);
 					},
 					function(next) {
+						if (!postData.toPid) {
+							return next(null);
+						}
+						db.sortedSetAdd('pid:' + postData.toPid + ':replies', timestamp, postData.pid, next);
```
Comment (barisusakli) [inline src/posts/create.js]: Can you change this to 

```
async.parallel([
    async.apply(db.sortedSetAdd , 'pid:' + postData.toPid + ':replies', timestamp, postData.pid),
    async.apply(db.incrObjectField, 'post:' + postData.toPid, 'replies')
]. next);
```

File: src/posts/delete.js [hunk 2]
```diff
@@ -140,6 +140,17 @@ module.exports = function(Posts) {
...
 					},
+					function(next) {
+						Posts.getPostField(pid, 'toPid', function (err, toPid) {
+							if (err) {
+								return next(err);
+							}
+							if (!parseInt(toPid, 10)) {
+								return next(null);
+							}
+							db.sortedSetRemove('pid:' + toPid + ':replies', pid, next);
```
Comment (barisusakli) [inline src/posts/delete.js]: Change this to 
```
async.parallel([
    async.apply(db.sortedSetRemove , 'pid:' + toPid + ':replies', pid),
    async.apply(db.decrObjectField, 'post:' + toPid, 'replies')
]. next);
```

File: src/topics/posts.js [hunk 3]
```diff
@@ -66,6 +66,9 @@ module.exports = function(Topics) {
 			voteData: function(next) {
 				favourites.getVoteStatusByPostIDs(pids, uid, next);
 			},
+			replies: function(next) {
+				posts.countReplies(pids, next);
```
Comment (barisusakli) [inline src/topics/posts.js]: remove this since `replies` will be stored in the post object

File: src/topics/posts.js [hunk 4]
```diff
@@ -124,6 +127,7 @@ module.exports = function(Topics) {
 					postObj.upvoted = results.voteData.upvotes[i];
 					postObj.downvoted = results.voteData.downvotes[i];
 					postObj.votes = postObj.votes || 0;
+					postObj.replies = results.replies[i] || 0;
```
Comment (barisusakli) [inline src/topics/posts.js]: Change this to `postObj.replies = postObj.replies || 0;`

File: src/upgrade.js [hunk 5]
```diff
@@ -10,7 +10,7 @@ var db = require('./database'),
 	schemaDate, thisSchemaDate,
 
 	// IMPORTANT: REMEMBER TO UPDATE VALUE OF latestSchema
-	latestSchema = Date.UTC(2016, 8, 22);
+	latestSchema = Date.UTC(2016, 8, 24);
```
Comment (barisusakli) [inline src/upgrade.js]: Update upgrade date

File: src/upgrade.js [hunk 6]
```diff
@@ -850,6 +850,41 @@ Upgrade.upgrade = function(callback) {
...
+					posts.getPostsFields(ids, ['pid', 'toPid', 'timestamp'], function(err, data) {
+						if (err) {
+							return next(err);
+						}
+
+						async.each(data, function(postData, next) {
+							if (!parseInt(post.toPid, 10)) {
+								return next(null);
+							}
+							db.sortedSetAdd('pid:' + postData.toPid + ':replies', postData.timestamp, postData.pid, next);
```
Comment (barisusakli) [inline src/upgrade.js]: Change this to

```
async.parallel([
    async.apply(db.sortedSetAdd , 'pid:' + postData.toPid + ':replies', postData.timestamp, postData.pid),
    async.apply(db.incrObjectField, 'post:' + postData.toPid, 'replies')
]. next);
```

File: src/posts.js [hunk 7]
```diff
@@ -258,4 +258,29 @@ var plugins = require('./plugins');
 		});
 	};
 
+	Posts.countReplies = function(pid, callback) {
```
Comment (barisusakli) [inline src/posts.js]: Can remove this function

**Comment by BenLubar** [comment]
@barisusakli

Ok, while testing this today (I'm planning to make the UI today), I got this error:

```
27/10 18:01 [31] - error: [upgrade] Errors were encountered while updating the NodeBB schema: E11000 duplicate key error index: 0.objects.$_key_1_value_-1 dup key: { : "post:26454", : null }
```

As far as I know, the only queries that ran that even touched something with that `_key` were:

```
posts.getPostsFields(ids, ['pid', 'toPid', 'timestamp'], function (err, data) {
```

which should be read-only and

```
async.apply(db.incrObjectField, 'post:' + postData.toPid, 'replies')
```

which shouldn't be inserting any new records into the database. In MongoDB console, `db.objects.findOne({_key: "post:26454"})` shows this:

```
{
        "_id" : ObjectId("56de5945f848f35aaeb980ad"),
        "_key" : "post:26454",
        "pid" : 26454,
        "uid" : 12,
        "tid" : 698,
        "content" : "@Alex Papadimoulis said:<blockquote>As far as the logo -- the reason it's red is because it's supposted to the result of a Rubber Stamp. Like, at a code review, someone would stamp a big WTF on your code. Get it?\r\n<p>BTW, would anyone ever buy those ruber stamps? I got a quote a while back for like $12/ea ... with shipping and whatnot probably could get them to people for $15 - $20 ea. seems a bit steep.</p>\r\n<p>&nbsp;</blockquote></p><p>Yeah. Makes sense. Nice idea.</p>Custom-made stamps are expensive, because they can't be mass-produced: you don't buy 1,000 stamps; you buy 2 or 3, and they last for a pretty long time. It involves a bit of manual labour and that's more costly than running the process through a big machine that spits out 10,000 stamps an hour.<br><br>Multi-colour staps are even more expensive -- basically they're several stamps layed out in one stamp... thing.<br><br>Red Bic pens, however, cost nothing at all. :)<br>",
        "timestamp" : 1147238664436,
        "votes" : 0,
        "editor" : "",
        "edited" : 0,
        "deleted" : 0,
        "ip" : "[redacted]",
        "_imported_pid" : 144112,
        "_imported_uid" : 16,
        "_imported_tid" : 5606,
        "_imported_content" : "[quote user=\"Alex Papadimoulis\"]As far as the logo -- the reason it's red is because it's supposted to the result of a Rubber Stamp. Like, at a code review, someone would stamp a big WTF on your code. Get it?\r\n<p>BTW, would anyone ever buy those ruber stamps? I got a quote a while back for like $12/ea ... with shipping and whatnot probably could get them to people for $15 - $20 ea. seems a bit steep.</p>\r\n<p>&nbsp;[/quote]</p><p>Yeah. Makes sense. Nice idea.</p>Custom-made stamps are expensive, because they can't be mass-produced: you don't buy 1,000 stamps; you buy 2 or 3, and they last for a pretty long time. It involves a bit of manual labour and that's more costly than running the process through a big machine that spits out 10,000 stamps an hour.<br><br>Multi-colour staps are even more expensive -- basically they're several stamps layed out in one stamp... thing.<br><br>Red Bic pens, however, cost nothing at all. :)<br>",
        "_imported_cid" : 13,
        "_imported_ip" : "[redacted]",
        "_imported_guest" : "",
        "_imported_toPid" : 71994,
        "_imported_user_slug" : "",
        "_imported_user_path" : "/users/dhromed",
        "_imported_topic_slug" : "",
        "_imported_topic_path" : "",
        "_imported_category_path" : "/c/general",
        "_imported_category_slug" : "general",
        "_imported_path" : ""
}
```

I'm restoring the staging server's database from a backup, but I have no idea how that error message could even correspond to the queries being run.

**Comment by barisusakli** [comment]
@BenLubar you can fix it by changing async.each to async.eachSeries

**Comment by barisusakli** [comment]
It sounds like we need to do this check in incrObjectField as well https://github.com/NodeBB/NodeBB/blob/master/src/database/mongo/sorted.js#L21-L23

**Comment by BenLubar** [comment]
Demonstration: https://youtu.be/3SDK9mGKGG4

**Comment by psychobunny** [comment]
Niiice. Needs a touch more love, let me know if I can help

**Comment by BenLubar** [comment]
@psychobunny if you have any ideas, go ahead!

**Comment by barisusakli** [comment]
This looks good, my only complaint is the icon is hardcoded to fa-plus, themes should be able to use something else if they want.

**Comment by barisusakli** [comment]
Something else I noticed, the OP also gets a `pid:<toPid>:replies`, do we need this? Since all posts are already displayed under the OP, i think it doesn't make sense to have it.

**Comment by BenLubar** [comment]
A post can be a reply directly to the OP or a reply to the topic in general or a reply to another post in the topic (or even in a different topic if one of the posts was moved).

**Comment by barisusakli** [comment]
@BenLubar I made some changes to the UI after the merge. Instead of using a different template I use the same post template from the topic.

image

**Pull request merged**